_posts/2023-10-11-kafka-consumer-client-essentials.md

```diff
@@ -0,0 +1,485 @@
+        props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, BOOTSTRAP_SERVERS);
+
+        // Set a unique client ID for tracking
+        props.put(ConsumerConfig.CLIENT_ID_CONFIG, "client-" + UUID.randomUUID());
+
+        //Set a consumer group ID for the consumer
+        props.put(ConsumerConfig.GROUP_ID_CONFIG, GROUP_ID);
+
+         // Configure deserializers for keys and values
+        props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, LongDeserializer.class);
+        props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class);
+
+        // Disable automatic offset committing
+        props.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, false);
+
+        //Set the offset reset behavior to start consuming from the earliest available offset
+        props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        return new KafkaConsumer<>(props);
+    }
+
+    private void sleep(long ms) {
+        try {
+            TimeUnit.MILLISECONDS.sleep(ms);
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private boolean retriable(Exception e) {
+        if (e == null) {
+            return false;
+        } else if (e instanceof IllegalArgumentException
+            || e instanceof UnsupportedOperationException
+            || !(e instanceof RebalanceInProgressException)
+            || !(e instanceof RetriableException)) {
+            return false;
+        } else {
+            return true;
+        }
+    }
+
+    @Override
+    public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
+        System.out.printf("Assigned partitions: %s%n", partitions);
+    }
+
+    @Override
+    public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
+        System.out.printf("Revoked partitions: %s%n", partitions);
+        kafkaConsumer.commitSync(pendingOffsets);
+        pendingOffsets.clear();
+    }
+    
+    @Override
+    public void onPartitionsLost(Collection<TopicPartition> partitions) {
+        System.out.printf("Lost partitions: {}", partitions);
+    }
+
+    @Override
+    public void onComplete(Map<TopicPartition, OffsetAndMetadata> map, Exception e) {
+        if (e != null) {
+            System.err.println("Failed to commit offsets");
+            if (!retriable(e)) {
+                e.printStackTrace();
+                System.exit(1);
+            }
+        }
+    }
+}
+```
+
+### Running the consumer application
+
+To put this client into action, simply run the main method in the `consumer` class. 
+The client consumes messages from the specified topic (`TOPIC_NAME`) until it reaches the predefined message count, which is 50 messages with the `NUM_MESSAGES` constant value we specified. 
+
+> The consumer is not designed to be safely accessed concurrently by multiple threads. Ensure that only one instance of the consumer application is running to maintain the expected behavior.
```

Comment:
Not thread safe means a single consumer instance should not be shared between multiple threads. You can run multiple consumer instances.